### DISCUSSION TITLE
Uploading file using FormData in Deno

### DISCUSSION THREAD

**Discussion by tillspeicher** [nhost/nhost#1851] [Q&A] [answered]
I'm trying to use Deno to upload files to Nhost Storage via the JS SDK.

I have some code which was able to do this successfully in ca. September or October 2022, but it looks like a bunch of things have changed since then, and I'm getting errors now.
The previous code looked as follows:
```typescript
async uploadFile(filePath: string, bucketId: string): Promise<string> {
    const fileName = path.basename(filePath)
    const fileContent = await Deno.readFile(filePath)
    const file = new File([fileContent], fileName)
    const { fileMetadata, error } = await this.nhost.storage.upload({ 
        file, name: fileName, bucketId,
    })
    //...
}
```

It looks like the suggested way to upload files now is via `FormData`, as described here: https://docs.nhost.io/reference/javascript/storage/upload
However, when I change the above function to
```typescript
async uploadFile(filePath: string, bucketId: string): Promise<string> {
    const fileName = path.basename(filePath)
    const fileData = new FormData()
    fileData.append("file", await Deno.readFile(filePath))
    const { fileMetadata, error } = await this.nhost.storage.upload({ 
        formData: fileData, name: fileName, bucketId,
    })
    //...
}
```
I get an error saying `{ message: "file[] not found in Multipart form" }`.

Judging from the Nhost Storage source code that parses the request (https://github.com/nhost/hasura-storage/blob/main/controller/upload_file.go#L198), it looks like a `file[]` key is expected, instead of a `file` key (though I'm not familiar with Go). Changing `.append("file",...)` to `.append("file[]",...)` also doesn't work, in any case.

Do you have an idea for how to fix this problem?
Is the issue related to passing a file object directly, instead of a using `fs.createReadStream(...)`? I haven't really found a way to create something equivalent to this call in Deno.

Thank you!

**Comment by szilarddoro** [comment]
Based on a previous conversation with one of our storage users, I created a small working demo. Please note that it's for Node, but the idea should be the same: https://github.com/szilarddoro/node-nhost-storage

My idea is that even though you are setting the form data correctly, what you are trying to set as value is actually empty.

**Comment by tillspeicher** [comment, marked as answer]
Thanks a lot for looking into the issue and posting the demo!

As you pointed out, the value I was trying to set in the `FormData` was actually empty.
I managed to fix the upload now by using the previous file object as the value:
```typescript
async uploadFile(filePath: string, bucketId: string): Promise<string> {
    const fileName = path.basename(filePath)
    const fileContent = await Deno.readFile(filePath)
    const file = new File([fileContent], fileName)
    const formData = new FormData()
    formData.append("file", file)
    const { fileMetadata, error } = await this.nhost.storage.upload({ 
        formData, name: fileName, bucketId,
    })
    //...
}
```
Thanks for pointing this out!

**Reply by szilarddoro** [reply]
I'm glad it worked! Let us know if there is anything else we can help you with.
